File: src/core/contracts/tree-evaluation.ts
import type { Call } from "../../engine/types";
import type { ConditionResult } from "./evidence-bundle";
import type { RecommendationBand } from "./meaning";

export interface SiblingConditionDetail extends Omit<ConditionResult, "satisfied"> {
  readonly name: string;
  readonly description: string;
}

/** Unified eligibility model — all dimensions that gate candidate selectability.
 *  CONSTRAINT: Pure DTO — no methods, no imports from conventions/ or engine/. */
export interface CandidateEligibility {
  readonly hand: {
    readonly satisfied: boolean;
    readonly failedConditions: readonly SiblingConditionDetail[];
  };
  readonly encoding: {
    readonly legal: boolean;
    readonly reason?: "all_encodings_illegal" | "illegal_in_auction";
  };
  readonly pedagogical: {
    readonly acceptable: boolean;
    readonly reasons: readonly string[];
  };
}

/** Strategy-resolved candidate — what the pipeline actually considered for this auction+hand.
 *  Serializable DTO (no function refs). */
export interface ResolvedCandidateDTO {
  readonly bidName: string;
  readonly meaning: string;
  readonly call: Call;
  readonly resolvedCall: Call;
  readonly isDefaultCall: boolean;
  readonly legal: boolean;
  readonly isMatched: boolean;
  readonly priority?: "preferred" | "alternative";
  readonly intentType: string;
  readonly failedConditions: readonly SiblingConditionDetail[];
  readonly eligibility?: CandidateEligibility;
  /** Ordering key for deterministic tie-breaking within selection tiers. */
  readonly orderKey?: number;
  /** All encoder encodings with legality — retained for teaching and obligation enrichment.
   *  `resolvedCall` is the first legal encoding (selection unchanged). */
  readonly allEncodings?: readonly { readonly call: Call; readonly legal: boolean }[];
  /** Originating module — threaded from MeaningProposal for downstream consumers. */
  readonly moduleId?: string;
  /** Semantic equivalence class — threaded from MeaningProposal. */
  readonly semanticClassId?: string;
  /** Authored recommendation band — threaded from MeaningProposal ranking. */
  readonly recommendationBand?: RecommendationBand;
}

/** Check eligibility dimensions for DTO — pedagogical is post-selection annotation, not a gate.
 *  Falls back to legacy `legal` + `failedConditions` when eligibility is absent. */
export function isDtoSelectable(c: ResolvedCandidateDTO): boolean {
  if (!c.eligibility) return c.legal && c.failedConditions.length === 0;
  return c.eligibility.hand.satisfied
    && c.eligibility.encoding.legal;
}

/** Check pedagogical acceptability on DTO — post-selection annotation for teaching consumers. */
export function isDtoTeachingAcceptable(c: ResolvedCandidateDTO): boolean {
  if (!c.eligibility) return true;
  return c.eligibility.pedagogical.acceptable;
}

// Re-export teaching/grading types that were moved to teaching-grading.ts.
// Kept here so existing consumers importing from tree-evaluation continue to work.
export type { AlternativeGroup, IntentFamily, IntentRelationship } from "./teaching-grading";

/** Structured trace of how the convention pipeline evaluated a bid.
 *  Always-on (not DEV-gated). Plain DTO — no convention-core imports. */
export interface EvaluationTrace {
  readonly conventionId: string;
  readonly candidateCount: number;
  /** True when a strategy result was rejected by the chain's resultFilter. */
  readonly forcingFiltered?: boolean;
  readonly strategyChainPath: readonly { readonly strategyId: string; readonly result: "suggested" | "declined" | "filtered" | "error" }[];
  /** Number of posterior samples drawn. Set by meaning-pipeline strategies with posterior engine. */
  readonly posteriorSampleCount?: number;
  /** Confidence from posterior engine (0-1). Set by meaning-pipeline strategies with posterior engine. */
  readonly posteriorConfidence?: number;
}

File: src-tauri/crates/bridge-conventions/src/adapter/tree_evaluation.rs
//! Candidate eligibility and resolved candidate DTO types.
//!
//! Mirrors TS types from `pipeline/tree-evaluation.ts`.

use bridge_engine::types::Call;
use serde::{Deserialize, Serialize};

use crate::types::meaning::RecommendationBand;

/// Detail about a sibling condition that failed.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct SiblingConditionDetail {
    pub name: String,
    pub description: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub condition_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub fact_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub observed_value: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub threshold: Option<serde_json::Value>,
}

/// Hand eligibility.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct HandEligibility {
    pub satisfied: bool,
    pub failed_conditions: Vec<SiblingConditionDetail>,
}

/// Encoding eligibility.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct EncodingEligibility {
    pub legal: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub reason: Option<EncodingFailureReason>,
}

/// Why encoding failed.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum EncodingFailureReason {
    AllEncodingsIllegal,
    IllegalInAuction,
}

/// Pedagogical eligibility.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct PedagogicalEligibility {
    pub acceptable: bool,
    pub reasons: Vec<String>,
}

/// Unified eligibility model — all dimensions that gate candidate selectability.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct CandidateEligibility {
    pub hand: HandEligibility,
    pub encoding: EncodingEligibility,
    pub pedagogical: PedagogicalEligibility,
}

/// Strategy-resolved candidate DTO.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ResolvedCandidateDTO {
    pub bid_name: String,
    pub meaning: String,
    pub call: Call,
    pub resolved_call: Call,
    pub is_default_call: bool,
    pub legal: bool,
    pub is_matched: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub priority: Option<CandidatePriority>,
    pub intent_type: String,
    pub failed_conditions: Vec<SiblingConditionDetail>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub eligibility: Option<CandidateEligibility>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub order_key: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub all_encodings: Option<Vec<EncodingOptionDTO>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub module_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub semantic_class_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub recommendation_band: Option<RecommendationBand>,
}

/// Candidate priority level.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum CandidatePriority {
    Preferred,
    Alternative,
}

/// Encoding option in resolved candidate.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct EncodingOptionDTO {
    pub call: Call,
    pub legal: bool,
}

/// Check eligibility dimensions for DTO — pedagogical is post-selection annotation, not a gate.
pub fn is_dto_selectable(c: &ResolvedCandidateDTO) -> bool {
    match &c.eligibility {
        Some(e) => e.hand.satisfied && e.encoding.legal,
        None => c.legal && c.failed_conditions.is_empty(),
    }
}

/// Check pedagogical acceptability on DTO.
pub fn is_dto_teaching_acceptable(c: &ResolvedCandidateDTO) -> bool {
    match &c.eligibility {
        Some(e) => e.pedagogical.acceptable,
        None => true,
    }
}
